Source item management

larsroettig opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
<!--- Provide a description of the changes proposed in the pull request -->

### Fixed Issues (if relevant)
<!--- Provide a list of fixed issues in the format magento/magento2#<issue_number>, if relevant  -->
1. magento/magento2#<issue_number>: Issue title
2. ...

### Manual testing scenarios
<!--- Provide a set of unambiguous steps to test the proposed code change -->
1. ...
2. ...

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

kieronthomas committed:
source item management

Commit:
MSI

- UI part of product assignment from source page

Commit:
Merge pull request #36 from jonfres/source-item-management

source item management

Commit:
MSI:

- Code Review Changes
- Implement List and Delete Function
- Implement Test for List and Delete Function

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem/Collection.php:
```diff
@@ -0,0 +1,53 @@
+            $connection,
+            $resource
+        );
+    }
```
do we really need to have this constructor if all that it does is just proxying input parameters to parent constructor

maghamed commented on app/code/Magento/Inventory/Model/SourceItem.php:
```diff
@@ -0,0 +1,76 @@
+    public function setStatus($status)
+    {
+        $this->setData(SourceItemInterface::STATUS, $status);
+    }
```
we also need to have next methods in this interface and implementation

getExtensionAttributes()/
setExtensionAttributes(\Magento\InventoryApi\Api\Data\SourceItemExtensionInterface $extensionAttributes)

larsroettig replied:
i fix this

maghamed commented on app/code/Magento/InventoryApi/Api/Data/SourceItemInterface.php:
```diff
@@ -0,0 +1,82 @@
+     * @param bool $status
+     * @return void
+     */
+    public function setStatus($status);
```
getExtensionAttributes()/
setExtensionAttributes(\Magento\InventoryApi\Api\Data\SourceItemExtensionInterface $extensionAttributes)

maghamed commented on app/code/Magento/InventoryApi/Api/Data/SourceItemInterface.php:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * Get source item status.
+     *
+     * @return bool
```
let's make status as INT
it will help us to reserve possible different statuses except "In STOCK" and "Out of Stock" if we will get some other statuses in future

maghamed commented on app/code/Magento/InventoryApi/Api/Data/SourceItemInterface.php:
```diff
@@ -0,0 +1,82 @@
+     * Constants for status value.
+     */
+    const IS_IN_STOCK = TRUE;
+    const IS_OUT_OF_STOCK = FALSE;
```
let's make values of these constants as 0 and 1

maghamed commented on app/code/Magento/InventoryApi/Api/SourceItemRepositoryInterface.php:
```diff
@@ -0,0 +1,45 @@
+     * on getList method, because entity loading way is different for these methods
+     *
+     * @param \Magento\InventoryApi\Api\Data\SourceItemInterface $sourceItem
+     * @return int
```
return type should be void, not int

maghamed commented on app/code/Magento/InventoryApi/Api/Data/SourceItemInterface.php:
```diff
@@ -0,0 +1,82 @@
+
+use Magento\Framework\Api\ExtensibleDataInterface;
+
+interface SourceItemInterface extends ExtensibleDataInterface
```
Let's add a description what's SourceItem interface represents. i.e. amount of particular product on some particular physical storage.

Also, let's add the description to the SourceInterface

maghamed commented on app/code/Magento/InventoryApi/Api/SourceItemRepositoryInterface.php:
```diff
@@ -0,0 +1,45 @@
+     * @return int
+     * @throws \Magento\Framework\Exception\CouldNotSaveException
+     */
+    public function save(\Magento\InventoryApi\Api\Data\SourceItemInterface $sourceItem);
```
let's make this method to accept an array of SourceItems in other case Repository::save would be performance bottleneck, because this method usually used for syncronization with ERP system and should be high performant. 

So, bulk operation should be supported

maghamed replied:
but it's not clear how to handle multiple statuses, we don't have dedicated interfaces which return statuses for Bulk API.
Like source item IDs 1,2,3,5 saved sucessfully
4 and 6 are failed to save.

Will raise this question on Architectural meeting

larsroettig replied:
@maghamed  but than we need some bulk save function the following code have also the same e  performance bottleneck. We can discuss this in an architectural call. 

```
 /**
     * @inheritdoc
     */
    public function save(array $sourceItemList)
    {
        try {
            
             $savedSourceItems = [];
            foreach ($sourceItemList as $sourceItem)
            {
                $this->resourceSource->save($sourceItem);
                $savedItems[] = $sourceItem->getSourceItemId();
            }
            
            return $savedSourceItems; 
        } catch (\Exception $e) {
            $this->logger->error($e->getMessage());
            throw new CouldNotSaveException(__('Could not save source item'), $e);
        }
    }
```

larsroettig replied:
I draf some posbile methode 
 ```
   /**
     * @param SourceItemInterface[] $sourceItemList
     * @return [] list of saved  
     */
    public function multiSave(array $sourceItemList)
    {


        $sourceItemData = [];
        foreach ($sourceItemList as $sourceItem) {
            $sourceItemData[] = [
                SourceItemInterface::SKU => $sourceItem->getSku(),
                SourceItemInterface::QUANTITY => $sourceItem->getQuantity(),
                /** @todo only a draft */
            ];
        }

        $connection = $this->connection->getConnection();
        $connection->insertMultiple(
            $connection->getTableName(InstallSchema::TABLE_NAME_SOURCE_ITEM),
            $sourceItemData
        );
        
        
    }
```

maghamed commented on app/code/Magento/InventoryApi/Api/SourceItemRepositoryInterface.php:
```diff
@@ -0,0 +1,45 @@
+     * @param \Magento\Framework\Api\SearchCriteriaInterface $searchCriteria
+     * @return \Magento\InventoryApi\Api\Data\SourceItemSearchResultsInterface
+     */
+    public function getList(\Magento\Framework\Api\SearchCriteriaInterface $searchCriteria = null);
```
this search criteria should not be nullable.

larsroettig replied:
why not for SourceRepository it is also nullable?

Commit:
MSI:
- Code Review Changes

Commit:
MSI: - Code Review Changes

Commit:
MSI:
- Code review changes
- Add new save methode for better performence
- Add new test case for save methode
- Remove nullable search SearchCriteria and the test case

Commit:
MSI

- Fixes after CR of Source item management #38

Commit:
Merge remote-tracking branch 'origin/source-item-management' into source-item-management

# Conflicts:
#	app/code/Magento/Inventory/Model/ResourceModel/SourceItem/Collection.php
#	app/code/Magento/InventoryApi/Api/Data/SourceItemInterface.php

maghamed commented on app/code/Magento/Inventory/Model/SourceItemRepository.php:
```diff
@@ -0,0 +1,163 @@
+
+            $connection->insertMultiple(
+                $connection->getTableName(InstallSchema::TABLE_NAME_SOURCE_ITEM),
+                $sourceItemData
```
insertMultiple - should be done by Resource model, not Repository.

The repository should be agnostic to the precise adapter which is used under the hood.
So, switching MySQL -> Postgres SQL -> Mongo DB should not affect Repository code. Just persistent layer should be affected (in our current conditions it's Resource model).

In your implementation Repository works directly with MySQL connection

maghamed commented on app/code/Magento/InventoryApi/Api/Data/SourceItemInterface.php:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * Set source item status.
+     *
+     * @param bool $status
```
status is not bool anymore

maghamed commented on app/code/Magento/InventoryApi/Api/Data/SourceItemInterface.php:
```diff
@@ -0,0 +1,102 @@
+     * Constants for status value.
+     */
+    const IS_IN_STOCK = 0;
+    const IS_OUT_OF_STOCK = 1;
```
as it's not bool, we could eliminate IS_ prefix

Commit:
MSI

- Source item management #38

Commit:
MSI - Pr38

- product form customization

Commit:
MSI - Pr38

- backend part

Commit:
MSI - Pr38

- backend part

Commit:
MSI - Pr38

- backend part

Commit:
MSI - Pr38

- backend part

maghamed commented on app/code/Magento/Inventory/Model/OptionSource/RegionSource.php:
```diff
@@ -10,6 +10,8 @@
 /**
  * Class RegionSource
+ *
+ * @api
```
why did we add api for these class?
should be removed

maghamed commented on app/code/Magento/Inventory/Model/OptionSource/CarrierSource.php:
```diff
@@ -10,6 +10,8 @@
 /**
  * Class CarrierSource
+ *
+ * @api
```
why did we add api for these class?
should be removed

maghamed commented on app/code/Magento/Inventory/Model/OptionSource/SourceItemStatus.php:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Class SourceItemStatus
+ *
+ * @api
```
Just interfaces should be marked as api. Not implementations

naydav replied:
It is source model

maghamed replied:
agree

Commit:
MSI - Pr38

- backend part

Commit:
MSI - Pr38

- fix unit tests

Commit:
MSI - Pr38

- fix static tests

Commit:
MSI - Pr38

- fix builds

naydav commented:
About failed tests:
1) https://travis-ci.org/magento-engcom/magento2/jobs/251955771
Will be fixed in https://github.com/magento-engcom/magento2/issues/44

2) app/code/Magento/Inventory/Observer/ProcessSourceItemsObserver.php:16	The class Magento\Inventory\Observer\ProcessSourceItemsObserver declared as final.
Will be discussed on internal Magento architecture meeting

maghamed commented on app/code/Magento/Inventory/Model/OptionSource/SourceItemStatus.php:
```diff
@@ -0,0 +1,40 @@
+     * Source items status values
+     */
+    const SOURCE_ITEM_STATUS_OUT_OF_STOCK = 0;
+    const SOURCE_ITEM_STATUS_IN_STOCK = 1;
```
why should we introduce new constants if we already have these constants declared in SourceItemInterface ?

naydav replied:
> We agreed that UNIQUE key would be added to SourceItem table (Source_id, SKU)
and that would be added in a separate module.


We have separate task for this
https://github.com/magento-engcom/magento2/issues/42

naydav replied:
> why should we introduce new constants if we already have these constants declared in SourceItemInterface ?

There are NOT new constants. Constants have been moved so Source model (like as in another source models)

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/Source/Collection.php:
```diff
@@ -16,6 +16,9 @@
 use Psr\Log\LoggerInterface;
 
+/**
+ * @api
```
if you add api PHP Doc Block , you must provide a description how this API supposed to be used

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem.php:
```diff
@@ -0,0 +1,56 @@
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+
+/**
+ * Class SourceItem
```
don't give useless comments

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem.php:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * @param Context $context
+     * @param MultipleSave $multipleSave
+     * @param null $connectionName
```
why do you have null in PHP Doc block ?

naydav replied:
Fixed

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem.php:
```diff
@@ -0,0 +1,56 @@
+     */
+    public function __construct(
+        Context $context,
+        MultipleSave $multipleSave,
```
please rename to MultiSave

naydav replied:
Renamed to SaveMultiple

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem.php:
```diff
@@ -0,0 +1,56 @@
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\Inventory\Model\ResourceModel;
```
why do we need to have this class at all , if all that it does is proxying call to 
Magento\Inventory\Model\ResourceModel\SourceItem\MultipleSave ?

naydav replied:
Fixed

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem/MultipleSave.php:
```diff
@@ -0,0 +1,117 @@
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+
+/**
+ * Class MultipleSave
```
don't provide useless description

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem/MultipleSave.php:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    /**
+     * Multiple save source items
```
Source Items multi save operation

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem/MultipleSave.php:
```diff
@@ -0,0 +1,117 @@
+    private $connection;
+
+    /**
+     * SourceCarrierLinkManagement constructor
```
why do you have constructor in the description?

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem/MultipleSave.php:
```diff
@@ -0,0 +1,117 @@
+     */
+    public function multipleSave(array $sourceItems)
+    {
+        if (!empty($sourceItems)) {
```
don't do conditional logic like this.
Make it like :  
```php
if (empty($sourceItems)) {
    return ;
}
```
doing so , it's much easier to read code of the method, because there is no unneeded offset.

naydav replied:
One point of function leave has better readability for me (and it is better for debugging)

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem/MultipleSave.php:
```diff
@@ -0,0 +1,117 @@
+     * @param array $bind
+     * @return string
+     */
+    private function buildValuesSqlPart(array $sourceItems, array &$bind)
```
it's not recommended to accept parameters by reference

maghamed replied:
please , rename this method

naydav replied:
Fixed

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem/MultipleSave.php:
```diff
@@ -0,0 +1,117 @@
+        return $sql;
+    }
+
+    /**
```
please provide description

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem/MultipleSave.php:
```diff
@@ -0,0 +1,117 @@
+     */
+    private function buildOnDuplicateSqlPart(array $fields)
+    {
+        $connection = $connection = $this->connection->getConnection();
```
why do you have two connection assignments?

naydav replied:
Fixed

maghamed commented on app/code/Magento/Inventory/Model/SourceItem.php:
```diff
@@ -0,0 +1,109 @@
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+
+/**
+ * Class SourceItem
```
don't give useless description

maghamed commented on app/code/Magento/Inventory/Model/SourceItemRepository.php:
```diff
@@ -0,0 +1,137 @@
+use Psr\Log\LoggerInterface;
+
+/**
+ * Class SourceItemRepository
```
please fix this

maghamed commented on app/code/Magento/Inventory/Observer/SourceItemsProcessor.php:
```diff
@@ -0,0 +1,158 @@
+use Magento\InventoryApi\Api\SourceItemSaveInterface;
+
+/**
+ * Class CarrierRequestDataHydrator
```
what?

maghamed commented on app/code/Magento/Inventory/Observer/SourceItemsProcessor.php:
```diff
@@ -0,0 +1,158 @@
+     * @param string $sku
+     * @return SourceItemInterface[]
+     */
+    private function getCurrentSourceItemsMap($sku)
```
source item map?
what's this?

naydav replied:
You can see this in method description
"Key is source id, value is Source Item"

maghamed commented on app/code/Magento/Inventory/Observer/SourceItemsProcessor.php:
```diff
@@ -0,0 +1,158 @@
+            unset($sourceItemsForDelete[$sourceId]);
+        }
+        $this->saveSourceItems($sourceItemsForSave);
+        $this->deleteSourceItems($sourceItemsForDelete);
```
looks like these operations should be atomic and processed in transaction

naydav replied:
As we discussed on Friday, we decide not use transation in this case

maghamed commented on app/code/Magento/Inventory/Observer/SourceItemsProcessor.php:
```diff
@@ -0,0 +1,158 @@
+        $sourceItems = $this->sourceItemRepository->getList($searchCriteria)->getItems();
+
+        $sourceItemMap = [];
+        if ($sourceItems) {
```
try to avoid unneeded offsets in code. It's harder to read code with offsets

maghamed commented on app/code/Magento/Inventory/Observer/SourceItemsProcessor.php:
```diff
@@ -0,0 +1,158 @@
+    private function validateSourceItemData(array $sourceItemData)
+    {
+        if (!isset($sourceItemData[SourceItemInterface::SOURCE_ID])) {
+            throw new InputException(__('Wrong Product to Source relation parameters.'));
```
parameters -> given

maghamed commented on app/code/Magento/Inventory/Observer/SourceItemsProcessor.php:
```diff
@@ -0,0 +1,158 @@
+     * @param array $sourceItems
+     * @return void
+     */
+    private function saveSourceItems(array $sourceItems)
```
why do we need to extract separate method for such simple and easy to read operation ?

maghamed commented on app/code/Magento/Inventory/Ui/DataProvider/Product/Form/Modifier/Sources.php:
```diff
@@ -0,0 +1,92 @@
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+
+/**
+ * Data provider that provide data about source items
```
please fix this description

naydav replied:
Fixed

maghamed commented on app/code/Magento/InventoryApi/Api/SourceItemSaveInterface.php:
```diff
@@ -0,0 +1,23 @@
+namespace Magento\InventoryApi\Api;
+
+/**
+ * Command for multiple saving of source items
```
fix the description

maghamed reviewed (changes requested):
We agreed that UNIQUE key would be added to SourceItem table (Source_id, SKU)
and that would be added in a separate module.
InventoryCatalog

naydav commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem/MultipleSave.php:
```diff
@@ -0,0 +1,117 @@
+     * @param array $bind
+     * @return string
+     */
+    private function buildValuesSqlPart(array $sourceItems, array &$bind)
```
> it's not recommended to accept parameters by reference

it conditioned by logic of method

naydav commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem.php:
```diff
@@ -0,0 +1,56 @@
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\Inventory\Model\ResourceModel;
```
Same logic as "proxies in repository"

Commit:
MSI - Pr38

- fixes after CR

Commit:
MSI - Pr38

- fixes after CR

Commit:
MSI - Pr38

- fixes after CR
